### ISSUE TITLE
Deprecate saturating_add_assign!() macro

### ISSUE THREAD

**Issue by steviez** [anza-xyz/agave#502]
#### Problem

The following convenience macro exists to perform add-assign with saturating math:
https://github.com/anza-xyz/agave/blob/9076348ef40e28975796c2ff25460bf3b25d41e4/sdk/src/lib.rs#L168-L175

However, support for this is stable in Rust as of 1.74.0:
https://doc.rust-lang.org/std/num/struct.Saturating.html#

#### Proposed Solution
- Updates all instances of `saturating_add_assign!()` macro to instead create `Saturating<T>` initially and then just use the `+=` operator
- Once all instances are removed, mark the macro as deprecated; it is a public function in `sdk` so we should probably let it ride as deprecated for a bit instead of immediately ripping out

**Comment by Nagaprasadvr** [comment]
fyi working on this

**Comment by Nagaprasadvr** [comment]
is this approach fine @steviez  ?

![image](https://github.com/anza-xyz/agave/assets/81755170/bf7ad0f4-ead2-4c6a-9340-eef43821eb8e)

**Comment by steviez** [comment]
> is this approach fine @steviez ?

Hi @Nagaprasadvr  - thank you for your interest in working on this. I think the code you included above would result in the values in `self` not actually getting updated (it looks like you're creating new variables instead of mutating the values in `self`). Something like https://github.com/anza-xyz/agave/pull/523 is how I envisioned it.

I'm going to get some feedback on my approach, and hopefully reach a decision with the direction we want to go. So, maybe hang tight until the PR I mentioned ships. Once we have a general approach we like, you can start knocking some more of these out if you're still interested

**Comment by Nagaprasadvr** [comment]
> > is this approach fine @steviez ?
> 
> Hi @Nagaprasadvr - thank you for your interest in working on this. I think the code you included above would result in the values in `self` not actually getting updated (it looks like you're creating new variables instead of mutating the values in `self`). Something like #523 is how I envisioned it.
> 
> I'm going to get some feedback on my approach, and hopefully reach a decision with the direction we want to go. So, maybe hang tight until the PR I mentioned ships. Once we have a general approach we like, you can start knocking some more of these out if you're still interested

sure you are right , i checked it with rust playground

**Comment by Nagaprasadvr** [comment]
tried this approach -
![image](https://github.com/anza-xyz/agave/assets/81755170/b219abfa-5711-470c-917c-394dc171af18)


Output - 
![image](https://github.com/anza-xyz/agave/assets/81755170/c24479bd-24e8-4db8-99c9-1e37229c9d66)

**Comment by steviez** [comment]
> Once all instances are removed, mark the macro as deprecated; it is a public function in sdk so we should probably let it ride as deprecated for a bit instead of immediately ripping out

The point of this issue is to get rid of the custom macro now that the standard library has a mechanism for what we want built in

**Comment by Nagaprasadvr** [comment]
> > Once all instances are removed, mark the macro as deprecated; it is a public function in sdk so we should probably let it ride as deprecated for a bit instead of immediately ripping out
> 
> The point of this issue is to get rid of the custom macro now that the standard library has a mechanism for what we want built in

ohk makes sense then we have to change types to `Saturating<T>`

**Comment by roryharr** [comment]
Looks like no one has looked at this for a long time. Is this something that is still needed?

I have some spare time and am looking into learning rust.

**Comment by steviez** [comment]
> Looks like no one has looked at this for a long time. Is this something that is still needed?
> 
> I have some spare time and am looking into learning rust.

This cleanup work is still desired so any PR's that reduce our usage (and lead to eventual deprecation) will be much appreciated. I had done some of these updates in an old branch, I'll see if I still have it around and get posted if so to help chip away at it

**Comment by Nagaprasadvr** [comment]
Will give it a try

**Comment by steviez** [comment]
@Nagaprasadvr - As an FYI, my comment was in response to @roryharr. If you're interested in looking at this again, I'd try to coordinate with roryharr as they said they plan on continuing this work [here](https://github.com/anza-xyz/agave/pull/3954#issuecomment-2524568930)

**Comment by Nagaprasadvr** [comment]
Cool!

**Comment by roryharr** [comment]
Yup, I plan on continuing this work, but many hands makes light work. 

To help avoid stepping on each others toes if we are both working on it, I've found starting from the data structure rather than focusing on the macro calls is the best way of approaching the issue.

I'll be working on all the data structures in the timings folder next. I'll likely have two PRs early next week, as there are a few warts in ProgramTiming to cleanup that might need more focused review.

**Comment by roryharr** [comment]
Looks like someone removed saturating from the timing library over the weekend. Looking at SVM now.

**Comment by roryharr** [comment]
Discussed during https://github.com/anza-xyz/agave/pull/4070 - It is not just a find and replace exercise. Saturating add is currently used in locations where it is not required. 

In general, each usage should be bucketed into three categories:
Overflow is not likely in production outside of a major bug: remove saturating_add_assign. replace with += as done in:
https://github.com/anza-xyz/agave/pull/4070

Overflow is possible and saturating is required: Replace with Saturating type.
Overflow is possible and saturating is not required: Replace with Wrapping type.

The default type (not Saturating or Wrapping) will panic in debug builds and wrap in production builds.

**Comment by roryharr** [comment]
Looking at leader_slot_metrics.rs, I would think the same applies as timing: They should just have the saturating math removed. They all get added through non saturating adds in consume_worker anyways. 


@apfitzge Does that sound correct?
